modulemd/include/modulemd-2.0/modulemd-module-stream.h

```diff
@@ -275,6 +278,25 @@ modulemd_module_stream_upgrade (ModulemdModuleStream *self,
                                 guint64 mdversion,
                                 GError **error);
 
+/**
+ * modulemd_module_stream_upgrade_ext:
+ * @self: (in): This #ModulemdModuleStream object.
+ * @mdversion: (in): The metadata version to upgrade to. If zero, upgrades to
+ * the highest-supported version.
+ * @error: (out): A #GError that will return the reason for an upgrade error.
+ *
+ * Return an upgraded copy of this object. Does not modify the original.
+ *
+ * Returns: (transfer full): A newly-allocated #ModulemdModuleIndex containing a
+ * copy of this object upgraded to the requested version. Returns NULL and sets
+ * @error appropriately if the upgrade could not be completed automatically.
+ *
+ * Since: 2.10
+ */
+ModulemdModuleIndex *
```

Comment:
@sgallagher It occurred to me that I can probably avoid some of the problems I was experiencing while trying to put my implementation of this function to use today by switching the return type to `ModulemdModule *`. It will avoid a lot of the overhead and internal complexity of using an index (such as it automatically attempting stream version upgrades at undesired times--which were causing me infinite recursion trouble). The automatic stream upgrades might even work. I'll give that a try tomorrow.